### ISSUE TITLE
Segmentation for measurement

### ISSUE THREAD

**Issue by IvanNut** [ultralytics/ultralytics#21622]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

There is a situation where I need to calculate the wall thickness of a sewer pipe using a camera. Is it possible? There is data for training the model, but is it worth the candle? In my head, the thought looks unreal, in the sense that the pipe is being shot in profile. Using the segmentation model, only the pipe edge is taken, and then it can be calculated. I hope this idea is real.

![Image](https://github.com/user-attachments/assets/a1327c85-9251-428c-a039-8af9f60293c3)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @IvanNut, thanks for opening this issue and for your interest in Ultralytics 🚀! This is an automated response to help you get started while an Ultralytics engineer reviews your question and will assist you soon.

We recommend a visit to the [Docs](https://docs.ultralytics.com/) for new users where you can find many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples and where many of the most common questions may already be answered.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (like segmentation-based measurement), please share additional details to help us guide you effectively:
-Goal: what exactly you want to measure (e.g., pipe wall thickness in pixels or real-world units) and acceptable error/variance📏
-Data: a few representative raw images and sample annotations/masks (inner/outer edges or wall region), plus info on how they were labeled
-Camera setup: distance to target, focal length/lens, resolution, lighting, whether lens distortion correction or calibration is applied
-Calibration: how you convert pixels to real-world units (e.g., known reference object or intrinsic/extrinsic calibration)
-Training details: YOLO11-seg model variant you plan to use, train/val split, image size, epochs, augmentations, and the exact commands run
-Logs: training logs and metrics (loss/precision/recall/seg metrics) and any console output
-Environment: OS, Python/PyTorch/CUDA versions, GPU model, and ultralytics package version
-Example workflow: how you plan to derive measurements from masks (e.g., extracting edges, post-processing)🧪
Also verify you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

Join the Ultralytics community where it suits you best. For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧. Prefer in-depth discussions? Check out [Discourse](https://community.ultralytics.com/). Or dive into threads on our [Subreddit](https://www.reddit.com/r/Ultralytics/) to share knowledge with the community.

## Upgrade

Upgrade to the latest `ultralytics` package including all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to verify your issue is not already resolved in the latest version:

```bash
pip install -U ultralytics
```

## Environments

YOLO may be run in any of the following up-to-date verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

**Comment by glenn-jocher** [comment]
Your idea is definitely feasible with YOLO11 instance segmentation! This approach can work well for measuring pipe wall thickness from profile images.

Here's how you can implement this:

1. **Custom train a YOLO11 segmentation model** on your pipe data to detect and segment the inner and outer pipe boundaries with pixel-level accuracy
2. **Extract boundary coordinates** from the segmentation masks to identify the pipe edges
3. **Calculate thickness** by measuring the distance between inner and outer boundaries at various points
4. **Apply calibration** to convert pixel measurements to real-world units using known reference dimensions

Since you have training data available, I'd recommend starting with a YOLO11 segmentation model and training it to segment both the inner and outer pipe surfaces. The [instance segmentation capabilities](https://www.ultralytics.com/blog/how-to-use-ultralytics-yolo11-for-instance-segmentation) will give you the precision needed for accurate measurements.

You can begin with:
```python
from ultralytics import YOLO

# Train custom model
model = YOLO('yolo11n-seg.pt')
model.train(data='your_pipe_dataset.yaml', epochs=100)

# Use for measurement
results = model('pipe_image.jpg')
# Extract segmentation masks and calculate thickness
```

This approach is commonly used in industrial inspection applications, so your concept is quite realistic and achievable.

**Comment by GANDA1F7** [comment]
Thank you for your guidance. This is very important to me.

**Comment by glenn-jocher** [comment]
Glad it helped—for measurement-grade results, calibrate and undistort the camera and include a known scale, keep the pipe as fronto-parallel as possible, then fit circles/ellipses to the inner and outer masks and use the radii difference; see the Instance Segmentation docs for training/val/export and the Instance Segmentation and Tracking guide for working with masks and videos.
